pkg/controller/sparkapplication/controller.go

```diff
@@ -523,17 +567,149 @@ func (c *Controller) updateSparkApplicationStatusWithRetries(
 	}
 
 	if lastUpdateErr != nil {
-		glog.Errorf("failed to update SparkApplication %s: %v", toUpdate.Name, lastUpdateErr)
+		glog.Errorf("failed to update SparkApplication %s: %v", original.Name, lastUpdateErr)
 	}
 
 	return nil
 }
 
+// isValidDriverStatusTransition determines if the driver state transition is valid
+func isValidDriverStatusTransition(oldStatus, newStatus v1alpha1.ApplicationStateType) bool {
+	for _, validStatus := range validDriverStateTransitions[oldStatus] {
+		if newStatus == validStatus {
+			return true
+		}
+	}
+	return false
+}
+
+// isValidExecutorStatusTransition determines if the executor state transition is valid
+func isValidExecutorStatusTransition(oldStatus, newStatus v1alpha1.ExecutorState) bool {
+
+	for _, validStatus := range validExecutorStateTransitions[oldStatus] {
+		if newStatus == validStatus {
+			return true
+		}
+	}
+
+	return false
+}
+
+func fetchMetricLabels(specLabels map[string]string, labels []string) map[string]string {
+
+	metricLabels := map[string]string{}
+
+	for _, label := range labels {
+		metricLabels[label] = specLabels[label]
+		if metricLabels[label] == "" {
+			metricLabels[label] = "Unknown"
+		}
+	}
+	return metricLabels
+}
+
+func (c *Controller) postMetrics(oldState, newState *v1alpha1.SparkApplication) {
+	if c.metrics == nil {
+		glog.V(2).Infof("Metrics not enabled. Skipping posting metrics")
```

Comment:
`Metrics not enabled so skipping exporting metrics`. The same below.

Reply:
done